### PR TITLE
MAM prepared queries WIP

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses "prepared queries for mam"

Proposed changes include:
* mod_mam_rdbms_arch is fully converted.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[8783.1](https://travis-ci.org/esl/MongooseIM/jobs/746252459) / Erlang 23.0.3 / small_tests / 8453762ff08f8a8e9cc1508f55cb473bd43eaa06
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2958/8783/small_tests.23.0.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2958/8783/small_tests.23.0.3/small/ct_run.mongooseim@localhost.2020-11-27_15.00.10/lib.mongooseim.logs/run.2020-11-27_15.00.10/suite.log.html)

File: src/mam/mod_mam_rdbms_arch.erl [hunk 1]
```diff
@@ -172,46 +187,40 @@ retract_message(Host, #{archive_id := UserID,
                         direction := Dir,
                         packet := Packet}) ->
     case mod_mam_utils:get_retract_id(mod_mam, Host, Packet) of
-        none -> ok;
-        OriginIDToRetract -> retract_message(Host, UserID, LocJID, RemJID, OriginIDToRetract, Dir)
+        none ->
+            ok;
+        OriginIDToRetract ->
+            retract_message(Host, UserID, LocJID, RemJID, OriginIDToRetract, Dir)
```
Comment (chrzaszcz) [inline src/mam/mod_mam_rdbms_arch.erl:191]: I don't see any improvement here. Why change it? I think it's OK to have one clause per line in simple cases. Easier to grep.

File: src/rdbms/rdbms_queries.erl [hunk 2]
```diff
@@ -823,20 +828,54 @@ create_bulk_insert_query(Table, Fields, RowsNum) when RowsNum > 0 ->
...
-    do_get_db_specific_limits(?RDBMS_TYPE, LimitStr).
+    do_get_db_specific_limits(?RDBMS_TYPE, LimitStr, false).
 
 -spec get_db_specific_offset(integer(), integer()) -> iolist().
 get_db_specific_offset(Offset, Limit) ->
     do_get_db_specific_offset(?RDBMS_TYPE, integer_to_list(Offset), integer_to_list(Limit)).
 
 
-do_get_db_specific_limits(mssql, LimitStr) ->
+do_get_db_specific_limits(mssql, LimitStr, false) ->
```
Comment (chrzaszcz) [inline src/rdbms/rdbms_queries.erl:838]: We definitely need a spec or a variable name here to understand what this 'false' and 'true' means.

File: src/mam/mod_mam_rdbms_arch.erl [hunk 3]
```diff
@@ -580,19 +535,14 @@ calc_offset(Host, F, _PS, _TC, #rsm_in{direction = aft, id = ID})
 calc_offset(_LS, _F, _PS, _TC, _RSM) ->
     0.
 
+maybe_reserve(asc, List) ->
+    List;
+maybe_reserve(desc, List) ->
+    lists:reverse(List).
+
```
Comment (chrzaszcz) [inline src/mam/mod_mam_rdbms_arch.erl]: Please, fix the name of this function.

**Review by chrzaszcz — COMMENTED** [review]
I just had a first look, not checking every detail yet. Could we change this PR so that it would be merged to a feature-branch? That branch might be created from master and when the whole MAM work is done (next sprint?) we would merge it all.

**Pull request closed without merging**